Find unique structured array elements more efficiently

Mysteryem opened the pull request:
np.unique works by sorting its input array and structured arrays are
sorted by sorting by each column sequentially with a stable sorting
algorithm. This means that every extra field in the structured array's
dtype adds an extra sort just to find unique elements.

When exporting shape key normals is enabled, 3 extra fields are added to
the `dots` array's structured dtype per shape key on the mesh, which can
quickly end up in over 100 sorts when finding the unique values.

By instead viewing a structured array as a non-structured dtype with the
same itemsize, unique values can be found with a single sort, but at the
cost of the result no longer being sorted according to the structured
dtype.

This non-structured dtype also has to be compared by its bytes, so -0.0
in any floating-point fields must be replaced with 0.0 beforehand and
any NaN values with different byte representations will be considered
unique from other NaN values.

For large meshes and/or meshes with many shape keys, this can reduce
export times by a few seconds.

---

This is heavily based on the similar [`io_scene_fbx.fbx_utils.fast_first_axis_unique`](https://projects.blender.org/blender/blender-addons/src/commit/a99528adb07d0c745b4f1ec5b1997e304b4dc048/io_scene_fbx/fbx_utils.py#L514) function I made for the FBX I/O add-on because finding unique elements along the first axis of a multidimensional array with np.unique internally just views each row and any extra dimensions as a structured dtype and then sorts that structured array view of the multidimensional array.

On a larger humanoid rigged model I have with a few hundred shape keys per mesh, this patch brings the export duration down from 40s to 31s (before this patch, it exports in 6s if I disable exporting shape key normals).

If desired, I can add explicit checks that the fields are not overlapping, though I think this is unlikely to ever be an issue.

If desired, I can add support for `bool` fields, though they don't appear to be used currently. For completeness, `bool` fields should set all values equal to `True` to `True`/`1` because its possible that `bool` fields could be constructed from a view of non-bool data. E.g. `np.arange(3, dtype=np.uint8).view(bool)` results in a `[False, True, True]` array, but the actual values will be `[0, 1, 2]`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/glTF-Blender-IO?pullRequest=2123) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/glTF-Blender-IO?pullRequest=2123) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/glTF-Blender-IO?pullRequest=2123) it.</sub>

julienduroure commented:
Thanks !
I will have a look soon

Mysteryem commented:
I missed a use of `np.unique` in `gltf2_blender_mesh.merge_duplicate_verts()`, so I have updated it to also use the new `fast_structured_np_unique` function.
I moved `fast_structured_np_unique` into `gltf2_blender_conversion` because it is no longer only used by `gltf_blender_gather_primitives_extract`.

julienduroure commented:
LGTM.
Maybe conversion file is not the best place for the function, as this is not a conversion.
Maybe we can create a _utils.py file?

julienduroure commented:
Merged! Thanks a lot
